### PR TITLE
Port cub::DeviceScanByKey tests to Catch2

### PR DESCRIPTION
**Pull request by elstehle**
## Description

This PR addresses https://github.com/NVIDIA/cccl/issues/85. It migrates tests for `DeviceReduceByKey` algorithms to Catch2. 

## Checklist

- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cub/test/catch2_test_device_scan.cuh [hunk 1]
```diff
@@ -56,3 +90,120 @@ void compute_inclusive_scan_reference(InputIt first,
     *result++ = acc;
```
Comment (senior-zero) [inline cub/test/catch2_test_device_scan.cuh:90]: important: this needs a cast to `OutputIt` value type

Comment (senior-zero) [reply cub/test/catch2_test_device_scan.cuh:90]: suggestion: we might want accumulator type in this function to be computed as `cub::detail::accumulator_t`

File: cub/test/catch2_test_device_scan_by_key_iterators.cu [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+  // Generate input segments
+  thrust::device_vector<offset_t> segment_offsets =
+    c2h::gen_uniform_offsets<offset_t>(CUB_SEED(1),
+                                       num_items,
+                                       std::get<0>(seg_size_range),
+                                       std::get<1>(seg_size_range));
+
+  // Get array of keys from segment offsets
+  const offset_t num_segments = static_cast<offset_t>(segment_offsets.size() - 1);
```
Comment (senior-zero) [inline cub/test/catch2_test_device_scan_by_key_iterators.cu:121]: important: unused, let's remove it to make msvc happy

File: cub/test/catch2_test_device_scan_by_key.cu [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+
+  // Generate input segments
+  thrust::device_vector<offset_t> segment_offsets =
+    c2h::gen_uniform_offsets<offset_t>(CUB_SEED(1),
+                                       num_items,
+                                       std::get<0>(seg_size_range),
+                                       std::get<1>(seg_size_range));
+
+  // Get array of keys from segment offsets
+  const offset_t num_segments = static_cast<offset_t>(segment_offsets.size() - 1);
```
Comment (senior-zero) [inline cub/test/catch2_test_device_scan_by_key.cu:121]: important: unused, let's remove to make msvc happy

File: cub/test/catch2_test_device_scan_by_key.cu [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
```
Comment (senior-zero) [inline cub/test/catch2_test_device_scan_by_key.cu:353]: important: unused, let's remove to make msvc happy

File: cub/test/catch2_test_device_scan.cuh (hunk 1 above)
Comment (senior-zero) [reply cub/test/catch2_test_device_scan.cuh:90]: important: same notes apply to `compute_exclusive_scan_reference`

File: cub/test/catch2_test_device_scan_by_key.cu (hunk 4 above)
Comment (elstehle) [reply cub/test/catch2_test_device_scan_by_key.cu]: Good catch, thanks 👍

File: cub/test/catch2_test_device_scan.cuh (hunk 1 above)
Comment (elstehle) [reply cub/test/catch2_test_device_scan.cuh]: Thanks! Added `cast`s and explicit `accumulator_t`. Feel free to give it a look, if you like 🙂

**Review by senior-zero — APPROVED** [review]

**Pull request merged**